In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
ecomdf=pd.read_excel('C:/Users/Acer/12JAN-18JAN2025-GENAI/DATASETS/E Commerce Dataset.xlsx',sheet_name='E Comm')

In [3]:
ecomdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   CustomerID                   5630 non-null   int64  
 1   Churn                        5630 non-null   int64  
 2   Tenure                       5366 non-null   float64
 3   PreferredLoginDevice         5630 non-null   object 
 4   CityTier                     5630 non-null   int64  
 5   WarehouseToHome              5379 non-null   float64
 6   PreferredPaymentMode         5630 non-null   object 
 7   Gender                       5630 non-null   object 
 8   HourSpendOnApp               5375 non-null   float64
 9   NumberOfDeviceRegistered     5630 non-null   int64  
 10  PreferedOrderCat             5630 non-null   object 
 11  SatisfactionScore            5630 non-null   int64  
 12  MaritalStatus                5630 non-null   object 
 13  NumberOfAddress   

 PREDICING NON NUMERIC VALUES

 RangeIndex : 5630 ( Observations )
 Columns : 20 variables
 Missing values in multiple variables ( columns )
 Data type - int,float,object

 variables that are categorical in nature like CityTier(1,2,3) , Complain ( 0/1 )
 Churn (0/1) - ( 0- No & 1 - Yes ) are int data type  
 Churn 1 - With the company , Churn 0 - Left the company


 Dependent Variable is Chrun(0/1)  - i.e. its a binary dependant variable


 BUILD A MACHINE LEARNING MODEL THAT PREDICTS PROBABILITY OF CHRUN OF EACH CUSTOMER





In [4]:
# Split data into numeric and objects

numcols=ecomdf.select_dtypes(include=np.number)
objcols=ecomdf.select_dtypes(include=['object'])


In [5]:
# Move column City TIer , Complain & Churn from numebric to objcols
objcols['CityTier']=numcols['CityTier']
objcols['Complain']=numcols['Complain']
objcols['Churn']=numcols['Churn']


In [6]:
numcols=numcols.drop(['CityTier','Complain','Churn'],axis=1)

In [7]:
numcols.describe()

CustomerID       Tenure  WarehouseToHome  HourSpendOnApp  \
count   5630.000000  5366.000000      5379.000000     5375.000000   
mean   52815.500000    10.189899        15.639896        2.931535   
std     1625.385339     8.557241         8.531475        0.721926   
min    50001.000000     0.000000         5.000000        0.000000   
25%    51408.250000     2.000000         9.000000        2.000000   
50%    52815.500000     9.000000        14.000000        3.000000   
75%    54222.750000    16.000000        20.000000        3.000000   
max    55630.000000    61.000000       127.000000        5.000000   

       NumberOfDeviceRegistered  SatisfactionScore  NumberOfAddress  \
count               5630.000000        5630.000000      5630.000000   
mean                   3.688988           3.066785         4.214032   
std                    1.023999           1.380194         2.583586   
min                    1.000000           1.000000         1.000000   
25%                    3.000000           2.000000         2.000000   
50%                    4.000000           3.000000         3.000000   
75%                    4.000000           4.000000         6.000000   
max                    6.000000           5.000000        22.000000   

       OrderAmountHikeFromlastYear   CouponUsed   OrderCount  \
count                  5365.000000  5374.000000  5372.000000   
mean                     15.707922     1.751023     3.008004   
std                       3.675485     1.894621     2.939680   
min                      11.000000     0.000000     1.000000   
25%                      13.000000     1.000000     1.000000   
50%                      15.000000     1.000000     2.000000   
75%                      18.000000     2.000000     3.000000   
max                      26.000000    16.000000    16.000000   

       DaySinceLastOrder  CashbackAmount  
count        5323.000000     5630.000000  
mean            4.543491      177.223030  
std             3.654433       49.207036  
min             0.000000        0.000000  
25%             2.000000      145.770000  
50%             3.000000      163.280000  
75%             7.000000      196.392500  
max            46.000000      324.990000

In [8]:
numcols.skew()

CustomerID                     0.000000
Tenure                         0.736513
WarehouseToHome                1.619154
HourSpendOnApp                -0.027213
NumberOfDeviceRegistered      -0.396969
SatisfactionScore             -0.142626
NumberOfAddress                1.088639
OrderAmountHikeFromlastYear    0.790785
CouponUsed                     2.545653
OrderCount                     2.196414
DaySinceLastOrder              1.191000
CashbackAmount                 1.149846
dtype: float64

In [9]:
# skew is closer to 0 hence we are using mean
# imputation of missing values using mean
 
for col in numcols.columns:
    numcols[col]=numcols[col].fillna(numcols[col].mean())

In [10]:
numcols.head()

CustomerID     Tenure  WarehouseToHome  HourSpendOnApp  \
0       50001   4.000000              6.0        3.000000   
1       50002  10.189899              8.0        3.000000   
2       50003  10.189899             30.0        2.000000   
3       50004   0.000000             15.0        2.000000   
4       50005   0.000000             12.0        2.931535   

   NumberOfDeviceRegistered  SatisfactionScore  NumberOfAddress  \
0                         3                  2                9   
1                         4                  3                7   
2                         4                  3                6   
3                         4                  5                8   
4                         3                  5                3   

   OrderAmountHikeFromlastYear  CouponUsed  OrderCount  DaySinceLastOrder  \
0                         11.0         1.0         1.0                5.0   
1                         15.0         0.0         1.0                0.0   
2                         14.0         0.0         1.0                3.0   
3                         23.0         0.0         1.0                3.0   
4                         11.0         1.0         1.0                3.0   

   CashbackAmount  
0          159.93  
1          120.90  
2          120.28  
3          134.07  
4          129.60

In [11]:
objcols.columns

Index(['PreferredLoginDevice', 'PreferredPaymentMode', 'Gender',
       'PreferedOrderCat', 'MaritalStatus', 'CityTier', 'Complain', 'Churn'],
      dtype='object')

In [12]:
#dummy encode objcols ( HOT ENCODING - convert the given columns in to binaries )
# i.e  exclude chrun

objcols_dummy=pd.get_dummies(objcols,columns=['PreferredLoginDevice', 'PreferredPaymentMode', 'Gender', 'PreferedOrderCat', 'MaritalStatus', 'CityTier', 'Complain'])

In [13]:
for col in objcols.columns:
    freq=objcols[col].value_counts(dropna=False) # It is used for showing missing values
    print(freq)

PreferredLoginDevice
Mobile Phone    2765
Computer        1634
Phone           1231
Name: count, dtype: int64
PreferredPaymentMode
Debit Card          2314
Credit Card         1501
E wallet             614
UPI                  414
COD                  365
CC                   273
Cash on Delivery     149
Name: count, dtype: int64
Gender
Male      3384
Female    2246
Name: count, dtype: int64
PreferedOrderCat
Laptop & Accessory    2050
Mobile Phone          1271
Fashion                826
Mobile                 809
Grocery                410
Others                 264
Name: count, dtype: int64
MaritalStatus
Married     2986
Single      1796
Divorced     848
Name: count, dtype: int64
CityTier
1    3666
3    1722
2     242
Name: count, dtype: int64
Complain
0    4026
1    1604
Name: count, dtype: int64
Churn
0    4682
1     948
Name: count, dtype: int64


In [14]:
objcols['PreferredPaymentMode']=objcols['PreferredPaymentMode'].str.replace("Cash on Delivery","COD")
objcols['PreferredPaymentMode']=objcols['PreferredPaymentMode'].str.replace("CC","Credit Card")
objcols['PreferedOrderCat']=objcols['PreferedOrderCat'].str.replace("Mobile Phone","Mobile")

In [15]:
#dummy encode objcols ( HOT ENCODING - convert the given columns in to binaries )
# i.e  exclude chrun

objcols_dummy=pd.get_dummies(objcols,columns=['PreferredLoginDevice', 'PreferredPaymentMode', 'Gender', 'PreferedOrderCat', 'MaritalStatus', 'CityTier', 'Complain'])

In [16]:
#concatenate both numeric cols and object cols
ecomdf_final=pd.concat([numcols,objcols_dummy],axis=1)

In [17]:
# split data into x and y

y=ecomdf_final['Churn']
X=ecomdf_final.drop(['Churn','CustomerID'],axis=1)


# Depenedent variable chrun is Binary Categorual variables (0/1 or No/ Yes )
# Classification Algorithms must be used


#BINARY LOGISTIC REGRESSION EXPLAINS RELATIONSHIP BETWEEN a binary dependant variable
#(y) and multiple independent variables (x)
#
#        e^B0+B1X1+B2X2+B3X3 ...... BnXn
# p = ---------------------------------
#    1+ e^B0+B1X1+B2X2+B3X3 ...... BnXn

# p - predicted probability that lies between 0 and 1
# e - exponential or 2.718
# B0 - intercept
# B1,B2,B3 .... Bn - Coefficients
# X1,X2,X3 .... Xn - Independent variables


For any classification we need to create confusion matrix

confusion matrix is a 2x2 Matrix of Actual Class & Predicted Class

Accuracy mist ne netweem .70 and .95

Accuracy-(TP+TN)/(TP+FN+FP+TN)

Other metrices are

Sensitivity or Recall= TP/ (TP+TN)  , i.r. how many positives are correctly predicted

Specificity = TN/(TN+FP ) . How many negative are correctly predicted as negative

Precision = TP/ (TP+FP). Overall Positive prediction rate

Alternative to accuracy is F1 score

                Precision * Recall
F1 Score= 2 X ----------------------
                Precision * Recall




![Alt Hi](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEiqxBWDxr82HWBNDD-XI22ht1rEZdhCUXilxAMtKFQugD4WxvEUUfqBv6fZfPwQ7MZZn_AM-xXGjhl2yeaHfWq47U2yKuigKcqy3wi5LQx7ExuJEnkdIJfyZHjpzRCkbYXil8z69kLf4a2_/s1600/confusionMatrxiUpdated.jpg)

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
logistic_model=LogisticRegression(max_iter=5000).fit(X,y)

In [21]:
logistic_model.score(X,y)

0.8966252220248668

In [22]:
logistic_pred=logistic_model.predict(X)

In [23]:
pd.crosstab(y,logistic_pred)
#True Positive - 4543 ; False Positive  - 139 ; False Positive - 443 ; True Negatives - 505

col_0     0    1
Churn           
0      4543  139
1       443  505

In [24]:
X.columns

Index(['Tenure', 'WarehouseToHome', 'HourSpendOnApp',
       'NumberOfDeviceRegistered', 'SatisfactionScore', 'NumberOfAddress',
       'OrderAmountHikeFromlastYear', 'CouponUsed', 'OrderCount',
       'DaySinceLastOrder', 'CashbackAmount', 'PreferredLoginDevice_Computer',
       'PreferredLoginDevice_Mobile Phone', 'PreferredLoginDevice_Phone',
       'PreferredPaymentMode_COD', 'PreferredPaymentMode_Credit Card',
       'PreferredPaymentMode_Debit Card', 'PreferredPaymentMode_E wallet',
       'PreferredPaymentMode_UPI', 'Gender_Female', 'Gender_Male',
       'PreferedOrderCat_Fashion', 'PreferedOrderCat_Grocery',
       'PreferedOrderCat_Laptop & Accessory', 'PreferedOrderCat_Mobile',
       'PreferedOrderCat_Others', 'MaritalStatus_Divorced',
       'MaritalStatus_Married', 'MaritalStatus_Single', 'CityTier_1',
       'CityTier_2', 'CityTier_3', 'Complain_0', 'Complain_1'],
      dtype='object')

In [25]:
X_new=X[['OrderAmountHikeFromlastYear','CouponUsed','OrderCount','DaySinceLastOrder','CashbackAmount']]

In [26]:
logit=LogisticRegression(solver="sag").fit(X_new,y)

In [27]:
print("Intercept",logit.intercept_)
print("coefficient",logit.coef_)

Intercept [0.0148851]
coefficient [[ 0.00542249  0.05737281  0.07678973 -0.1486106  -0.00846536]]


In [28]:
logit.predict_log_proba

<bound method LogisticRegression.predict_log_proba of LogisticRegression(solver='sag')>

# ######################## UPDATE ON 15 JAN 2025 ##########################

Included code 

In [29]:
from sklearn.tree import DecisionTreeClassifier ,plot_tree
from sklearn.model_selection import cross_val_score
#import pandas as pd
#import numpy as np
#import matplotlib.pyplot as plt
#%matplotlib inline

In [30]:
treemodel=DecisionTreeClassifier().fit(X,y)

In [31]:
treemodel.score(X,y)

1.0

In [33]:
cross_val_score(treemodel,X,y,cv=5)

array([0.98579041, 0.97513321, 0.98934281, 0.97690941, 0.98756661])

In [34]:
pd.DataFrame(X.columns,treemodel.feature_importances_).sort_index(ascending=False)

0
0.274901                               Tenure
0.087273                      WarehouseToHome
0.077438                    DaySinceLastOrder
0.072472                       CashbackAmount
0.066544                      NumberOfAddress
0.057719          OrderAmountHikeFromlastYear
0.046905                           Complain_1
0.038771             NumberOfDeviceRegistered
0.032754                 MaritalStatus_Single
0.031960                    SatisfactionScore
0.017696                           Complain_0
0.017414             PreferredPaymentMode_COD
0.016630     PreferredPaymentMode_Credit Card
0.016094                           OrderCount
0.015362                       HourSpendOnApp
0.015027        PreferredLoginDevice_Computer
0.014227                           CityTier_3
0.012455      PreferredPaymentMode_Debit Card
0.012070        PreferredPaymentMode_E wallet
0.011705  PreferedOrderCat_Laptop & Accessory
0.009511                           CityTier_1
0.009240           PreferredLoginDevice_Phone
0.008426              PreferedOrderCat_Mobile
0.007609                           CityTier_2
0.006603             PreferedOrderCat_Fashion
0.006440                MaritalStatus_Married
0.004888                        Gender_Female
0.004464                          Gender_Male
0.002903                           CouponUsed
0.002602             PreferredPaymentMode_UPI
0.001896              PreferedOrderCat_Others
0.000000               MaritalStatus_Divorced
0.000000             PreferedOrderCat_Grocery
0.000000    PreferredLoginDevice_Mobile Phone

# get the variables used by cross validation Score .As these are greater than 0